File: NetDeck/NRDB.m
//
//  NRDB.m
//  Net Deck
//
//  Created by Gereon Steffens on 20.06.14.
//  Copyright (c) 2015 Gereon Steffens. All rights reserved.
//

@import SDCAlertView;
@import AFNetworking;

#import "EXTScope.h"
#import "NRDB.h"
#import "NRDBAuth.h"
#import "SettingsKeys.h"

@interface NRDB()
@property (strong) DecklistCompletionBlock decklistCompletionBlock;
@property (strong) SaveCompletionBlock saveCompletionBlock;
@property NSMutableDictionary* deckMap;
@property NSTimer* timer;
@end

@implementation NRDB

static NRDB* instance;
static NSDateFormatter* formatter;

+(NRDB*) sharedInstance
{
    if (!instance)
    {
        instance = [[NRDB alloc] init];
    }
    return instance;
}

+(void) initialize
{
    formatter = [[NSDateFormatter alloc] init];
    [formatter setDateFormat:@"yyyy'-'MM'-'dd'T'HH':'mm':'ss'Z'"];
    formatter.timeZone = [NSTimeZone timeZoneWithName:@"GMT"];
}

+(void) clearSettings
{
    NSUserDefaults* settings = [NSUserDefaults standardUserDefaults];
    
    [settings removeObjectForKey:NRDB_ACCESS_TOKEN];
    [settings removeObjectForKey:NRDB_REFRESH_TOKEN];
    [settings removeObjectForKey:NRDB_TOKEN_EXPIRY];
    [settings removeObjectForKey:NRDB_TOKEN_TTL];
    [settings setObject:@(NO) forKey:USE_NRDB];
    
    [settings synchronize];
    [[NRDB sharedInstance].timer invalidate];
    [[UIApplication sharedApplication] setMinimumBackgroundFetchInterval:UIApplicationBackgroundFetchIntervalNever];
}

#pragma mark autorization

-(void) authorizeWithCode:(NSString *)code completion:(AuthCompletionBlock)completionBlock
{
    // NSLog(@"auth code");
    // ?client_id=" CLIENT_ID "&client_secret=" CLIENT_SECRET "&grant_type=authorization_code&redirect_uri=" CLIENT_HOST "&code="
    
    NSDictionary* parameters = @{
        @"client_id": @CLIENT_ID,
        @"client_secret": @CLIENT_SECRET,
        @"grant_type": @"authorization_code",
        @"redirect_uri": @CLIENT_HOST,
        @"code": code
    };
    
    [self getAuthorization:parameters completion:completionBlock];
}

-(void) refreshToken:(AuthCompletionBlock)completionBlock
{
    // NSLog(@"refresh token");
    // ?client_id=" CLIENT_ID "&client_secret=" CLIENT_SECRET "&grant_type=refresh_token&redirect_uri=" CLIENT_HOST "&refresh_token="

    NSString* token = [[NSUserDefaults standardUserDefaults] stringForKey:NRDB_REFRESH_TOKEN];
    
    if (token == nil)
    {
        completionBlock(NO);
        return;
    }
    
    NSDictionary* parameters = @{
        @"client_id": @CLIENT_ID,
        @"client_secret":  @CLIENT_SECRET,
        @"grant_type": @"refresh_token",
        @"redirect_uri": @CLIENT_HOST,
        @"refresh_token": token
    };

    [self getAuthorization:parameters completion:completionBlock];
}

-(void) getAuthorization:(NSDictionary*)parameters completion:(AuthCompletionBlock)completionBlock
{
    BOOL foreground = [UIApplication sharedApplication].applicationState == UIApplicationStateActive;
    if (foreground && !APP_ONLINE)
    {
        completionBlock(NO);
        return;
    }
    
    AFHTTPRequestOperationManager* manager = [AFHTTPRequestOperationManager manager];
    manager.responseSerializer = [AFJSONResponseSerializer serializer];
    
    [manager GET:@TOKEN_URL parameters:parameters
         success:^(AFHTTPRequestOperation* operation, id responseObject) {
             // NSLog(@"auth response: %@", responseObject);
             BOOL ok = YES;
             
             NSUserDefaults* settings = [NSUserDefaults standardUserDefaults];
             NSString* token = responseObject[@"access_token"];
             if (token)
             {
                 [settings setObject:token forKey:NRDB_ACCESS_TOKEN];
             }
             else
             {
                 ok = NO;
             }
             
             token = responseObject[@"refresh_token"];
             if (token)
             {
                 [settings setObject:token forKey:NRDB_REFRESH_TOKEN];
             }
             else
             {
                 ok = NO;
             }
             NSNumber* exp = responseObject[@"expires_in"];
             [settings setObject:exp forKey:NRDB_TOKEN_TTL];
             NSDate* expiry = [[NSDate date] dateByAddingTimeInterval:[exp intValue]];
             [settings setObject:expiry forKey:NRDB_TOKEN_EXPIRY];
             
             if (!ok)
             {
                 [NRDB clearSettings];
             }
             [settings synchronize];
             
             // NSLog(@"nrdb (re)auth success, status: %d", ok);
             [[UIApplication sharedApplication] setMinimumBackgroundFetchInterval:BG_FETCH_INTERVAL];
             completionBlock(ok);
         }
         failure:^(AFHTTPRequestOperation* operation, NSError* error) {
             [NRDB clearSettings];
             
             // NSLog(@"nrdb (re)auth failed: %@", operation);
             [SDCAlertView alertWithTitle:nil
                                  message:l10n(@"Authorization at NetrunnerDB.com failed")
                                  buttons:@[ l10n(@"OK") ]];
             
             [[UIApplication sharedApplication] setMinimumBackgroundFetchInterval:UIApplicationBackgroundFetchIntervalNever];
             completionBlock(NO);
         }
    ];
}

-(void) refreshAuthentication
{
    NSUserDefaults* settings = [NSUserDefaults standardUserDefaults];
    
    if (![settings boolForKey:USE_NRDB] || !APP_ONLINE)
    {
        return;
    }
    
    if ([settings objectForKey:NRDB_REFRESH_TOKEN] == nil)
    {
        [NRDB clearSettings];
        return;
    }
    
    NSDate* expiry = [settings objectForKey:NRDB_TOKEN_EXPIRY];
    NSDate* now = [NSDate date];
    NSTimeInterval diff = [expiry timeIntervalSinceDate:now];
    diff -= 5*60; // 5 minutes overlap
    // NSLog(@"start nrdb auth refresh in %f seconds", diff);

    if (diff < 0)
    {
        // token is expired, refresh now
        [self timedRefresh:nil];
    }
    else
    {
        self.timer = [NSTimer scheduledTimerWithTimeInterval:diff
                                                      target:self
                                                    selector:@selector(timedRefresh:)
                                                    userInfo:nil
                                                     repeats:NO];
    }
}

-(void) backgroundRefreshAuthentication:(BackgroundFetchCompletionBlock)completionHandler
{
    // NSLog(@"start bg fetch");
    
    NSUserDefaults* settings = [NSUserDefaults standardUserDefaults];
    if (![settings boolForKey:USE_NRDB])
    {
        // NSLog(@"no nrdb account");
        completionHandler(UIBackgroundFetchResultNoData);
        return;
    }
    
    if ([settings objectForKey:NRDB_REFRESH_TOKEN] == nil)
    {
        // NSLog(@"no token");
        [NRDB clearSettings];
        completionHandler(UIBackgroundFetchResultNoData);
        return;
    }

    [self refreshToken:^(BOOL ok) {
        // NSLog(@"refresh: %d", ok);
        completionHandler(ok ? UIBackgroundFetchResultNewData : UIBackgroundFetchResultFailed);
    }];
}

-(void) timedRefresh:(NSTimer*) timer
{
    [self refreshToken:^(BOOL ok) {
        NSTimeInterval ti = 300;
        if (ok)
        {
            NSNumber* ttl = [[NSUserDefaults standardUserDefaults] objectForKey:NRDB_TOKEN_TTL];
            ti = [ttl floatValue];
            ti -= 300; // 5 minutes before expiry
            
        }
        self.timer = [NSTimer scheduledTimerWithTimeInterval:ti
                                                      target:self
                                                    selector:@selector(timedRefresh:)
                                                    userInfo:nil
                                                     repeats:NO];
    }];
}

-(void) stopRefresh
{
    // NSLog(@"stopping refresh timer");
    [self.timer invalidate];
    self.timer = nil;
}

#pragma mark deck list

-(void) decklist:(DecklistCompletionBlock)completionBlock
{
    self.decklistCompletionBlock = completionBlock;
    [self getDecks];
}

-(void) getDecks
{
    NSString* decksUrl = @"http://netrunnerdb.com/api_oauth2/decks";
    
    NSString* accessToken = [[NSUserDefaults standardUserDefaults] stringForKey:NRDB_ACCESS_TOKEN];
    NSError* error;
    NSDictionary* params = @{
        @"access_token" : accessToken ? accessToken : @""
    };
    NSMutableURLRequest *request = [[AFHTTPRequestSerializer serializer] requestWithMethod:@"GET"
                                                                                 URLString:decksUrl
                                                                                parameters:params
                                                                                     error:&error];
    // bypass cache!
    request.cachePolicy = NSURLRequestReloadIgnoringLocalCacheData;
    
    AFHTTPRequestOperation *operation = [[AFHTTPRequestOperation alloc] initWithRequest:request];
    operation.responseSerializer = [AFJSONResponseSerializer serializer];
    
    @weakify(self);
    [operation setCompletionBlockWithSuccess:^(AFHTTPRequestOperation *operation, id responseObject) {
        @strongify(self);
        [self finishedDecklist:YES decks:responseObject];
    }
    failure:^(AFHTTPRequestOperation *operation, NSError *error) {
        @strongify(self);
        [self finishedDecklist:NO decks:nil];
    }];
    [operation start];
}

-(void) finishedDecklist:(BOOL)ok decks:(NSArray<Deck*>*)decks
{
    if (!ok)
    {
        [SDCAlertView alertWithTitle:nil
                             message:l10n(@"Loading decks from NetrunnerDB.com failed")
                             buttons:@[l10n(@"OK")]];
    }
    
    self.decklistCompletionBlock(decks);
}

#pragma mark publish deck

-(void) publishDeck:(Deck *)deck completion:(SaveCompletionBlock)completionBlock
{
    self.saveCompletionBlock = completionBlock;
    [self publishDeck:deck];
}

-(void) publishDeck:(Deck *)deck
{
    NSString* publishUrl = [NSString stringWithFormat:@"http://netrunnerdb.com/api_oauth2/publish_deck/%@", deck.netrunnerDbId];
    
    NSString* accessToken = [[NSUserDefaults standardUserDefaults] stringForKey:NRDB_ACCESS_TOKEN];
    NSDictionary* parameters = @{
        @"access_token" : accessToken ? accessToken : @""
    };

    [self saveOrPublish:publishUrl parameters:parameters];
}

#pragma mark load deck

-(void) loadDeck:(Deck *)deck completion:(LoadCompletionBlock)completionBlock
{
    NSString* accessToken = [[NSUserDefaults standardUserDefaults] stringForKey:NRDB_ACCESS_TOKEN];
    if (!accessToken)
    {
        completionBlock(NO, nil);
        return;
    }

    NSAssert(deck.netrunnerDbId, @"no nrdb id");
    NSString* loadUrl = [NSString stringWithFormat:@"http://netrunnerdb.com/api_oauth2/load_deck/%@", deck.netrunnerDbId];
    
    NSDictionary* parameters = @{ @"access_token" : accessToken };
    NSError* error;
    NSMutableURLRequest *request = [[AFHTTPRequestSerializer serializer] requestWithMethod:@"GET"
                                                                                 URLString:loadUrl
                                                                                parameters:parameters
                                                                                     error:&error];
    // bypass cache!
    request.cachePolicy = NSURLRequestReloadIgnoringLocalCacheData;
    
    AFHTTPRequestOperation *operation = [[AFHTTPRequestOperation alloc] initWithRequest:request];
    operation.responseSerializer = [AFJSONResponseSerializer serializer];
    
    [operation setCompletionBlockWithSuccess:^(AFHTTPRequestOperation *operation, id responseObject) {
        Deck* deck = [self parseDeckFromJson:responseObject];
        completionBlock(YES, deck);
    }
    failure:^(AFHTTPRequestOperation *operation, NSError *error) {
        // NSLog(@"save failed: %@", operation);
        completionBlock(NO, nil);
    }];
    [operation start];
}

-(Deck*) parseDeckFromJson:(NSDictionary*)json
{
    Deck* deck = [[Deck alloc] init];
    
    deck.name = json[@"name"];
    deck.notes = json[@"description"];
    deck.tags = json[@"tags"];
    deck.netrunnerDbId = [NSString stringWithFormat:@"%ld", (long)[json[@"id"] integerValue]];
    
    // parse last update '2014-06-19T13:52:24Z'
    deck.lastModified = [formatter dateFromString:json[@"dateupdate"]];
    deck.dateCreated = [formatter dateFromString:json[@"datecreation"]];
    
    for (NSDictionary* c in json[@"cards"])
    {
        NSString* code = c[@"card_code"];
        NSNumber* qty = c[@"qty"];
        
        Card* card = [CardManager cardByCode:code];
        if (card && qty)
        {
            [deck addCard:card copies:qty.intValue history:NO];
        }
    }
    
    NSArray* history = json[@"history"];
    
    NSMutableArray* revisions = [NSMutableArray array];

    for (NSDictionary* dict in history)
    {
        NSString* datecreation = dict[@"datecreation"];
        // NSLog(@"changeset created: %@", datecreation);
        DeckChangeSet* dcs = [[DeckChangeSet alloc] init];
        dcs.timestamp = [formatter dateFromString:datecreation];
        
        NSArray* variation = dict[@"variation"];
        NSAssert(variation.count == 2, @"wrong variation count");
        // 2-element array: variation[0] contains additions, variation[1] contains deletions
        
        for (int i=0; i<variation.count; ++i)
        {
            NSDictionary* dict = variation[i];
            
            // skip over empty and non-dictionary entries
            if (dict.count == 0) // || ![dict isKindOfClass:[NSDictionary class]])
            {
                continue;
            }
            
            for (NSString* code in [dict allKeys])
            {
                NSNumber* quantity = dict[code];
                NSInteger qty = quantity.integerValue;
                if (i == 1)
                {
                    qty = -qty;
                }
                
                Card* card = [CardManager cardByCode:code];
                
                if (card && qty)
                {
                    [dcs addCardCode:card.code copies:qty];
                }
            }
        }
        [dcs sort];
        [revisions addObject:dcs];
    }
    
    DeckChangeSet* initial = [[DeckChangeSet alloc] init];
    initial.initial = YES;
    initial.timestamp = deck.dateCreated;
    [revisions addObject:initial];
    
    deck.revisions = revisions;
    
    DeckChangeSet* newest = deck.revisions[0];
    NSMutableDictionary* cards = [NSMutableDictionary dictionary];
    for (CardCounter* cc in deck.allCards)
    {
        cards[cc.card.code] = @(cc.count);
    }
    newest.cards = [NSMutableDictionary dictionaryWithDictionary:cards];
    
    // walk through the deck's history and pre-compute a card list for every revision
    for (int i = 0; i < deck.revisions.count-1; ++i)
    {
        DeckChangeSet* prev = deck.revisions[i];
        for (DeckChange* dc in prev.changes)
        {
            NSNumber* qty = cards[dc.code];
            qty = @(qty.integerValue - dc.count);
            if (qty.integerValue == 0)
            {
                [cards removeObjectForKey:dc.code];
            }
            else
            {
                cards[dc.code] = qty;
            }
        }
        DeckChangeSet* dcs = deck.revisions[i+1];
        dcs.cards = [NSMutableDictionary dictionaryWithDictionary:cards];
    }
    
    /*
    for (int i=0; i < deck.revisions.count; ++i)
    {
        DeckChangeSet* dcs = deck.revisions[i];
        NSLog(@"%d %@", i, dcs.cards);
    }
    */
    
    return deck;
}

#pragma mark save deck

-(void) saveDeck:(Deck*)deck completion:(SaveCompletionBlock)completionBlock
{
    self.saveCompletionBlock = completionBlock;
    
    NSString* accessToken = [[NSUserDefaults standardUserDefaults] stringForKey:NRDB_ACCESS_TOKEN];
    if (accessToken)
    {
        [self saveDeck:deck];
    }
    else
    {
        completionBlock(NO, nil);
    }
}

-(void) saveDeck:(Deck *)deck
{
    NSMutableArray* json = [NSMutableArray array];
    if (deck.identity)
    {
        [json addObject:@{ @"card_code": deck.identity.code, @"qty": @1 }];
    }
    for (CardCounter* cc in deck.cards)
    {
        [json addObject:@{ @"card_code": cc.card.code, @"qty": @(cc.count) }];
    }
    NSData* jsonData = [NSJSONSerialization dataWithJSONObject:json options:NSJSONWritingPrettyPrinted error:nil];
    NSString* jsonStr = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];

    NSString* deckId = @"0";
    if (deck.netrunnerDbId)
    {
        deckId = deck.netrunnerDbId;
    }
    NSString* saveUrl = [NSString stringWithFormat:@"http://netrunnerdb.com/api_oauth2/save_deck/%@", deckId];
    
    NSString* accessToken = [[NSUserDefaults standardUserDefaults] stringForKey:NRDB_ACCESS_TOKEN];
    NSMutableDictionary* parameters = [NSMutableDictionary dictionary];
    parameters[@"access_token"] = accessToken ? accessToken : @"";
    parameters[@"content"] = jsonStr;
    if (deck.name)
    {
        parameters[@"name"] = deck.name;
    }
    if (deck.notes)
    {
        parameters[@"description"] = deck.notes;
    }
    if (deck.netrunnerDbId)
    {
        parameters[@"id"] = deck.netrunnerDbId;
    }
    if (deck.tags.count > 0)
    {
        parameters[@"tags"] = [deck.tags componentsJoinedByString:@" "];
    }
    
    [self saveOrPublish:saveUrl parameters:parameters];
}

-(void) saveOrPublish:(NSString*)url parameters:(NSDictionary*)parameters
{
    NSError* error;
    NSMutableURLRequest *request = [[AFHTTPRequestSerializer serializer] requestWithMethod:@"GET"
                                                                                 URLString:url
                                                                                parameters:parameters
                                                                                     error:&error];
    
    AFHTTPRequestOperation *operation = [[AFHTTPRequestOperation alloc] initWithRequest:request];
    operation.responseSerializer = [AFJSONResponseSerializer serializer];
    
    @weakify(self);
    [operation setCompletionBlockWithSuccess:^(AFHTTPRequestOperation *operation, id responseObject) {
        @strongify(self);
        BOOL success = [responseObject[@"success"] boolValue];
        if (success)
        {
            id idFromJson = responseObject[@"message"][@"id"];
            
            NSString* deckId;
            if ([idFromJson isKindOfClass:[NSString class]])
            {
                deckId = idFromJson;
            }
            else if ([idFromJson isKindOfClass:[NSNumber class]])
            {
                deckId = ((NSNumber*)idFromJson).stringValue;
            }
            if (deckId)
            {
                self.saveCompletionBlock(YES, deckId);
            }
            else
            {
                self.saveCompletionBlock(NO, nil);
            }
        }
        else
        {
            self.saveCompletionBlock(NO, nil);
        }
    }
    failure:^(AFHTTPRequestOperation *operation, NSError *error) {
        @strongify(self);
        // NSLog(@"save failed: %@", operation);
        self.saveCompletionBlock(NO, nil);
    }];
    [operation start];
}

#pragma mark deck map

-(void) updateDeckMap:(NSArray<Deck*>*)decks
{
    self.deckMap = [NSMutableDictionary dictionary];
    for (Deck* deck in decks)
    {
        if (deck.netrunnerDbId)
        {
            [self.deckMap setObject:deck.filename forKey:deck.netrunnerDbId];
        }
    }
}

-(void) deleteDeck:(NSString*) deckId
{
    if (deckId)
    {
        [self.deckMap removeObjectForKey:deckId];
    }
}

-(NSString*) filenameForId:(NSString*)deckId
{
    return [self.deckMap objectForKey:deckId];
}

@end

File: NetDeck/NRDB.swift
//
//  NRDB.swift
//  NetDeck
//
//  Created by Gereon Steffens on 07.03.16.
//  Copyright © 2021 Gereon Steffens. All rights reserved.
//

import Alamofire
import SwiftyUserDefaults

class NRDB: NSObject {
    static let clientHost = "netdeck://oauth2"
    static let clientId = "4_1onrqq7q82w0ow4scww84sw4k004g8cososcg8gog004s4gs08"
    static let clientSecret = "2myhr1ijml6o4kc0wgsww040o8cc84oso80o0w0s44k4k0c84"
    static let providerHost = "https://netrunnerdb.com"
    
    static let authUrl = providerHost + "/oauth/v2/auth?client_id=" + clientId + "&response_type=code&redirect_uri=" + clientHost
    static let tokenUrl = providerHost + "/oauth/v2/token"
    
    static let fiveMinutes: TimeInterval = 300 // in seconds
    
    static let sharedInstance = NRDB()
    override private init() {}
    
    private var timer: Timer?
    private var deckMap = [String: String]()
    
    static func clearSettings(setUseNrdb: Bool) {
        if setUseNrdb {
            Defaults[.useNrdb] = false
        }
        Defaults[.nrdbLoggedin] = false
        
        Defaults.remove(.nrdbAccessToken)
        Defaults.remove(.nrdbRefreshToken)
        Defaults.remove(.nrdbTokenExpiry)
        Defaults.remove(.nrdbTokenTTL)
        
        NRDB.sharedInstance.stopAuthorizationRefresh()
    }
    
    // MARK: - authorization
    
    func authorizeWithCode(_ code: String, completion: @escaping (Bool, String) -> Void) {
        // print("NRDB authWithCode")
        let parameters = [
            "client_id": NRDB.clientId,
            "client_secret": NRDB.clientSecret,
            "grant_type": "authorization_code",
            "redirect_uri": NRDB.clientHost,
            "code": code
        ]
        
        DispatchQueue.main.async {
            self.getAuthorization(parameters, isRefresh: false, completion: completion)
        }
    }
    
    private func refreshToken(_ completion: @escaping (Bool, String) -> Void) {
        // print("NRDB refreshToken")
        
        guard let token = Defaults[.nrdbRefreshToken] else {
            completion(false, "no refresh token")
            return
        }
        
        let formatter = DateFormatter()
        formatter.dateStyle = .long
        Defaults[.lastRefresh] = formatter.string(from: Date())
        
        let parameters = [
            "client_id": NRDB.clientId,
            "client_secret":  NRDB.clientSecret,
            "grant_type": "refresh_token",
            "redirect_uri": NRDB.clientHost,
            "refresh_token": token
        ]
        
        DispatchQueue.main.async {
            self.getAuthorization(parameters, isRefresh: true, completion: completion)
        }
    }
    
    private func getAuthorization(_ parameters: [String: String], isRefresh: Bool, completion: @escaping (Bool, String) -> Void) {
        // print("NRDB get Auth")
        let foreground = UIApplication.shared.applicationState == .active
        if foreground && !Reachability.online {
            completion(false, "offline")
            return
        }
        
        Alamofire.request(NRDB.tokenUrl, parameters: parameters)
            .validate()
            .responseJSON { response in
                switch response.result {
                case .success:
                    if let data = response.data {
                        var ok = true
                        do {
                            let decoder = JSONDecoder()
                            let auth = try decoder.decode(NetrunnerDbAuth.self, from: data)
                            
                            Defaults[.nrdbAccessToken] = auth.accessToken
                            Defaults[.nrdbRefreshToken] = auth.refreshToken
                            Defaults[.nrdbTokenTTL] = auth.expiresIn
                            let expiry = Date(timeIntervalSinceNow: auth.expiresIn)
                            Defaults[.nrdbTokenExpiry] = expiry
                        } catch let error {
                            print("auth error: bad json: \(error)")
                            ok = false
                        }
                        if ok {
                            if !Defaults[.keepNrdbCredentials] {
                                UIApplication.shared.setMinimumBackgroundFetchInterval(UIApplication.backgroundFetchIntervalMinimum)
                            }
                            completion(ok, "")
                        } else {
                            self.handleAuthorizationFailure(isRefresh, "bad json", completion: completion)
                        }
                    } else {
                        print("auth error: no data")
                    }
                case .failure:
                    print("auth error \(String(describing: response.response?.statusCode))")
                    if let data = response.data {
                        let body = String(data: data, encoding: .utf8)
                        print("body: \(String(describing: body))")
                    }
                    self.handleAuthorizationFailure(isRefresh, "http or server error", completion: completion)
                }
            }
    }
    
    private func handleAuthorizationFailure(_ isRefresh: Bool, _ error: String, completion: (Bool, String) -> Void) {
        NRDB.clearSettings(setUseNrdb: false)
        if !isRefresh {
            UIAlertController.alert(withTitle: nil, message: "Authorization at NetrunnerDB.com failed".localized(), button: "OK")
            UIApplication.shared.setMinimumBackgroundFetchInterval(UIApplication.backgroundFetchIntervalNever)
            completion(false, error)
            return
        }

        if Defaults[.keepNrdbCredentials] {
            NRDBHack.sharedInstance.silentlyLogin()
        }
    }
    
    func startAuthorizationRefresh() {
        // print("NRDB startAuthRefresh timer=\(self.timer)")
        
        if !Defaults[.useNrdb] || self.timer != nil {
            return
        }
        
        if Defaults[.nrdbRefreshToken] == nil {
            return
        }
        
        if NRDBHack.sharedInstance.loggingIn {
            return
        }
        
        let expiry = Defaults[.nrdbTokenExpiry] ?? Date()
        let now = Date()
        let diff = expiry.timeIntervalSince(now) - NRDB.fiveMinutes
        // print("start nrdb auth refresh in \(diff) seconds");
        
        self.timer?.invalidate()
        self.timer = Timer.scheduledTimer(timeInterval: diff, target: self, selector: #selector(NRDB.timedRefresh(_:)), userInfo: nil, repeats: false)
    }
    
    func stopAuthorizationRefresh() {
        // print("NRDB stopAuthRefresh")
        self.timer?.invalidate()
        self.timer = nil
        UIApplication.shared.setMinimumBackgroundFetchInterval(UIApplication.backgroundFetchIntervalNever)
    }
    
    @objc func timedRefresh(_ timer: Timer?) {
        // print("NRDB refresh timer callback")
        self.timer = nil
        self.refreshToken { ok, error in
            if ok {
                // schedule next run at 5 minutes before expiry
                let ti = Defaults[.nrdbTokenTTL] - NRDB.fiveMinutes
                // print("next refresh in \(ti) seconds")
                self.timer = Timer.scheduledTimer(timeInterval: ti, target: self, selector: #selector(NRDB.timedRefresh(_:)), userInfo: nil, repeats: false)
            }
        }
    }
    
    func backgroundRefreshAuthentication(_ completion: @escaping (UIBackgroundFetchResult) -> Void) {
        // print("NRDB background Refresh")
        if !Defaults[.useNrdb] {
            print("no nrdb account");
            completion(.noData)
            return
        }
        
        if Defaults[.nrdbRefreshToken] == "" {
            // print("no token")
            completion(.noData)
            return
        }
        
        self.refreshToken { ok, error in
            print("refresh ok \(ok) \(error)")
            completion(ok ? .newData : .failed)
        }
    }
    

    private func accessToken() -> String? {
        return Defaults[.nrdbAccessToken]
    }
    
    // MARK: - deck lists
    
    func decklist(_ completion: @escaping ([Deck]?) -> Void) {
        let accessToken = self.accessToken() ?? ""
        let decksUrl = URL(string: "https://netrunnerdb.com/api/2.0/private/decks?access_token=" + accessToken)!
    
        let request = URLRequest(url: decksUrl, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 10)

        Alamofire.request(request).validate().responseJSON { response in
            var decks: [Deck]?
            switch response.result {
            case .success:
                if let data = response.data {
                    decks = Deck.arrayFromJson(data)
                    completion(decks)
                }
            case .failure:
                completion(nil)
            }
            
            if decks == nil {
                UIAlertController.alert(withTitle: nil, message:"Loading decks from NetrunnerDB.com failed".localized(), button:"OK".localized())
            }
        }
    }
    
    func loadDeck(_ deckId: String, completion: @escaping (Deck?) -> Void) {
        guard let accessToken = self.accessToken() else {
            completion(nil)
            return
        }
        
        let loadUrl = URL(string: "https://netrunnerdb.com/api/2.0/private/deck/" + deckId + "?include_history=1&access_token=" + accessToken)!
        
        let request = URLRequest(url: loadUrl, cachePolicy: .reloadIgnoringLocalCacheData, timeoutInterval: 10)
        
        Alamofire.request(request).validate().responseJSON { response in
            switch response.result {
            case .success:
                if let data = response.data {
                    let decks = Deck.arrayFromJson(data)
                    completion(decks.first)
                }
            case .failure:
                completion(nil)
            }
        }
    }
    
    // MARK: - save / publish
    func saveDeck(_ deck: Deck, completion: @escaping (Bool, String?) -> Void) {
        guard let accessToken = self.accessToken() else {
            completion(false, nil)
            return
        }
        
        var cards = [String: Int]()
        if let id = deck.identity {
            cards[id.code] = 1
        }
        for cc in deck.cards {
            cards[cc.card.code] = cc.count
        }
        
        let saveUrl = "https://netrunnerdb.com/api/2.0/private/deck/save?access_token=" + accessToken
        let deckId = deck.netrunnerDbId ?? "0"
        let parameters: [String: Any] = [
            "deck_id": deckId,
            "name": deck.name,
            "description": deck.notes ?? "",
            "content": cards
        ]
        
        self.saveOrPublish(saveUrl, parameters: parameters, completion: completion)
    }
    
    func publishDeck(_ deck: Deck, completion: @escaping (Bool, String?) -> Void) {
        guard let accessToken = self.accessToken() else {
            completion(false, nil)
            return
        }
        
        let publishUrl = "https://netrunnerdb.com/api/2.0/private/deck/publish?access_token=" + accessToken
        
        let parameters = [
            "deck_id": deck.netrunnerDbId ?? "0",
            "name": deck.name
        ]
        
        self.saveOrPublish(publishUrl, parameters:parameters, completion: completion)
    }

    func saveOrPublish(_ url: String, parameters: [String: Any], completion: @escaping (Bool, String?)->Void) {
        Alamofire.request(url, method: .post, parameters: parameters, encoding: JSONEncoding.default)
            .validate()
            .responseJSON { response in
                switch response.result {
                case .success:
                    if let data = response.data {
                        let decks = Deck.arrayFromJson(data)
                        if decks.count > 0 {
                            if let deck = decks.first, deck.netrunnerDbId != "" {
                                completion(true, deck.netrunnerDbId)
                                return
                            }
                        } else {
                            completion(false, nil)
                            return
                        }
                    }
                case .failure:
                    break
                }
                completion(false, nil)
            }
    }
    
    // MARK: - mapping of nrdb ids to filenames
    
    func updateDeckMap(_ decks: [Deck]) {
        self.deckMap.removeAll()
        for deck in decks {
            self.addDeck(deck)
        }
    }
    
    func addDeck(_ deck: Deck) {
        if let filename = deck.filename, let nrdbId = deck.netrunnerDbId {
            self.deckMap[nrdbId] = filename
        }
    }
    
    func filenameForId(_ deckId: String?) -> String? {
        return self.deckMap[deckId ?? ""]
    }
    
    func deleteDeck(_ deckId: String?) {
        self.deckMap.removeValue(forKey: deckId ?? "")
    }

}
